bundles/org.openhab.binding.bmwconnecteddrive/README.md

```diff
@@ -0,0 +1,862 @@
+
+There are many Channels available for each Vehicle. 
+For better overview they are clustered in different Channel Groups.
+The Channel Groups are different for the Vehicle Types, on the build-in Sensors of the Vehicle and the activated Services.
+
+### Bridge Channels
+
+If the Vehicle isn't found in the Discovery force a log of Fingerprint Data which helps to analyze the problem.
+Please check [TroubleShooting Section](#TroubleShooting) for further advice.
+
+| Channel Label         | Channel ID            | Type   | Description                                       |
+|-----------------------|-----------------------|--------|---------------------------------------------------|
+| Discovery Fingerprint | discovery-fingerprint | Switch | Forcing a log entry to analyze Discovery Problems |
+
+### Thing Channel Groups 
+
+#### Vehicle Status
+
+Reflects Status of the Vehicle.
+Available for all Vehicles, Read-only.
+
+| Channel Label             | Channel Group ID | Channel ID          | Type          | Description                                                                       |
+|---------------------------|------------------|---------------------|---------------|-----------------------------------------------------------------------------------|
+| Overall Door Status       | status           | doors               | String        | **Closed** if all closed otherwise **Open**. **Unknown** if no data is delivered  |
+| Overall Window Status     | status           | windows             | String        | **Closed** if all closed otherwise **Open** or **Intermediate**. **Unknown** if no data is delivered   |
+| Doors Locked              | status           | lock                | String        | Status if Doors are locked or unlocked                                            |
+| Next Service Date         | status           | service-date        | DateTime      | Date of Upcoming Service                                                          |
+| Mileage till Next Service | status           | service-mileage     | Number:Length | Mileage till Upcoming Service                                                     |
+| Check Control             | status           | check-control       | String        | Indicator if CheckControl is **Active** or **Not Active**. **Unknown** if no data is delivered         |
+| Charging Status           | status           | charge              | String        | Only available for PHEV, BEV_REX and BEV                                          |
+| Last Status Timestamp     | status           | last-update         | DateTime      | Date and Time of last status update.                                              |
+
+See [further details for DateTime](#last-status-update-timestamp) in case of wrong timestamp values
+
+#### Services
+
+Group for all upcoming Services with Description, Service Date and / or Service Mileage
+It's designed as [List Interface](#list-interface) so all upcoming Services can be checked.
```

Comment:
The concept used is the [Sliding Window](https://www.youtube.com/watch?v=9-3BXsfrpbY). Window size is 1 so only one value is shown in UI. It's common in Automotive products e.g. showing contact list or music tracks on the infotainment display with e.g. 5 entries visible. Currently there's no default UI widget which applies to this Interface - unfortunately I didn't found any good UI list representation so far. 

For Rules I don't get the problem. If a rule selects a track on a music player the actual track number & description, track time, cover art, and so on changes in UI. Same as selecting another CheckControl Message.

The intended approach isn't clear to me how to solve some basic use cases. Perhaps I'm missing something:
In principle the car is ok and doesn't show any CheckControl message. If e.g. the wiper fluid runs dry a corresponding message is raised. I'll create a new channel and update the data - from binding view I'm fine! But no item is attached to observe changes in rules or in UI. Same happens if another message is raised on top. Here I'm stucked!

I don't think this problem applies especially to this binding. I see a more general flaw handling dynamic content. Or I didn't understood the whole story so far.

Reply:
@weymann I understand your issue here, but I also agree with @fwolter that this is mixing in UI/navigation functionality, which does not belong into a binding.
A binding should provide information from the remote device as distinct channels to consume - the information should not change based on where the user just navigates such a list (what happens if two users do that at the same time?). So @fwolter's suggestion with multiple channels would fit better to the general concept of how to model things in a binding. But you definitely have a point that those dynamic channels would be rather useless as there won't be any items attached to them.

The best suggestion I can give is to resort to Thing actions: Those are to be used when requests/commands need parameters or when the result is a complex data type that cannot be mapped to a single item type.
You could create a `Service getService(int index)` method that could be used within rules. If the user wants to have such information in the UI, he can implement what he needs (maybe he wants only the dates of the 3 next services displayed) and fill appropriate items with that.
Furthermore, it would allow full access to the list from rules, which makes it a much more flexible solution.

Reply:
The goal was simply to present the user all possible values. If there are 3 warnings it has to be possible to select them all. Bindings which I personally use I see different approaches solving (likely) the same list problem

- caldav binding is using eventNr in config to get the next n appointments
- openweathermap binding is using [channel groups to list the weatherforcast data](https://www.openhab.org/addons/bindings/openweathermap/#_3-hour-forecast). For 3-hour forecast these are 40 Channel Groups with each 19 Channel Id's and sums up to 760 items!

I'm going now with the solution from LG webOS: A String stores the current selected TV-Channel and all other possibilities are attached as _options_ using the _BaseDynamicStateDescriptionProvider_. I've seen in OH3 these options are provided in th UI so for me it should do the trick for now.

Reply:
The new solution you came up with is a bit of an improvement, but has quite the same flaws as the last: The user writes to a Channel which causes other Channels to change, but that doesn't actually change any state of the Thing. Channels are bound to a Thing, e.g. if you change the program on a TV, the state of the Thing is actually changed and thus the Channels.

Did you take a look into [Thing Actions](https://www.openhab.org/docs/developer/bindings/#actions-bound-to-a-thing) as @kaikreuzer suggested?

Reply:
Yes, I've taken a look into the Action suggestion. With this the complete list can be delivered and checked in rules. Then the user can decide what to do and introduce _n-items_ to be shown in the UI. This seems to me like another implementation option like creating dynamic channels - it's just shifted to the user.

Honestly I don't think the item update without state change is a real flaw. Even in rules you're able to update items with _postUpdate_ without changing device state. So hopefully this isn't forbidden in the binding.

The only flaw I see is the parallel usage if several users/rules are sending commands. **But this problem is already present** in other locations. I've air conditioning via MELCloud. The [operationMode channel](https://www.openhab.org/addons/bindings/melcloud/#channels) is providing the same interface - String with options.
Assume the temperature is 21 C and

* My wife wants to _Heat_
* I want to _Cool_
* Rule wants to _Dry_ - humidity too high

If this happens in parallel simply the last one wins. So either me, my wife or even both of us are wondering why the state isn't the desired one. Of course this is a different use case with _real device control_ instead of _browsing_ but the underlying problem is the same.

Reply:
This is a tough topic, because the UI doesn't support to show lists (in fact it's a table) and the question is now to apply a hack or let the user deal with it.

What about this? Taking the upcoming services as an example. Combine the events in one String Channel:

ServiceName1: ServiceDate1/MileageTillService1, ServiceName2: ServiceDate2/MileageTillService2, ...

If no services are upcoming the Channel would be e.g. "No services upcoming".

Reply:
For sure it's a tough topic, but it's worth discussing it :)

I still don't think it's a hack. If we look simply at the user interface it's the same as the LGWebOS binding where I get the idea from. You select a new TV-channel and corresponding channels are changing. Under the "technical hood" there's a real TV state change, for BMW it's _just browsing_. 

Allow me to present one use case:
It's the Habpanel with _Destination Selection Item_ and _OpenStreetMap_ Widget. The binding with the current pushed implementation is used.

1. Select a different destination address
2. The item with GPS coordinates changes
3. The new coordinates will be shown as new marker in the map

https://user-images.githubusercontent.com/20279271/103712963-66f8eb80-4fbb-11eb-88c4-7ad82d7ba626.mp4

No rules needed
Just used the standard Selection widget
All _last destinations_ are present in the UI and can be selected

### Rules
Setting my OH3 environment I just found out that you can get all data in rules with 
```
item.stateDescription.options
```
This is exactly what @kaikreuzer mentioned in his first suggestion, to make all data accessible in rules. So basically you can access everything in rules without interacting with the UI! This wasn't clear to me before.

### Finally ...
I found the binding with exactly the same problem. It's the [carnet binding](https://github.com/openhab/openhab-addons/pull/9091 ) and unfortunately you're one of the reviewers :) 
They provide _last destinations_ - same as this binding. Check [readme ](https://github.com/markus7017/openhab-addons/tree/carnet_snapshot3/bundles/org.openhab.binding.carnet) how it's implemented:

1. In thing configuration you need to configure how many shall be available 0-5
2. For each entry a new ChannelGroup _destination**x**_ is introduced
3. Each group contains 8 Channels

So up to 40 Channels are provided for the last destinations. Don't misunderstand me, the solution isn't wrong! Nevertheless setup isn't really easy, maybe there are more last destinations available or even less than you configured.

Reply:
Maybe @markus7017 can give a short impression how he uses the num-options of his binding in the UI.

Reply:
Can you comment on my proposal?

> Combine the events in one String Channel:
> 
> ServiceName1: ServiceDate1/MileageTillService1, ServiceName2: ServiceDate2/MileageTillService2, ...
> 
> If no services are upcoming the Channel would be e.g. "No services upcoming".

Reply:
For sure it's possible doing it that way. It's also possible to write in the json directly. With this the a parser is needed e.g. to get next service date or mileage in an own value.

But I don't know which problem will be solved with this pattern.  As I stated in my previous comment: I don't see a hack anymore.

Reply:
The usability of your current implementation as shown in the video looks good, but the original problem remains.

> If we look simply at the user interface it's the same as the LGWebOS binding where I get the idea from. You select a new TV-channel and corresponding channels are changing. Under the "technical hood" there's a real TV state change, for BMW it's just browsing.

Your last sentence describes the problem I see very well. In your example, switching of the TV channel changes an actual state of the Thing, but browsing through the services doesn't change the Thing's state.

If two users browse through the services, they would interfere each other as they change the state simultaneously. That issue exists with all Things, but in this case the user doesn't expect an interference as he/she is only browsing through a list in the UI.

I don't think you need a parser if you provide the servies via an Action for machine reading and a human readable string for the presentation in the UI.

Reply:
I fully agree. That's the only remaining problem.

> That issue exists with all Things, but in this case the user doesn't expect an interference as he/she is only browsing through a list in the UI. 

Please take a minute to think about it  ... this is the only impact! The binding doesn't crash, no data is inconsistent and no functional issues. It's _only_ two users in _the same household_ who may observe this in _rare occasions_. 

For me it's valid to take this constraint and mention it e.g. in readme instead of  creating a complete different concept with _list in one String_ for UI and _Actions_ for rules which makes provide the shown _Map use case_ impossible / more difficult.

Reply:
@openhab/add-ons-maintainers Could please anybody join this discussion? We're some kind of stuck here and need some other opinion.

Reply:
I don't want to block any progress, so I'm okay with sticking to the current solution.

Reply:
> I still don't think it's a hack. If we look simply at the user interface it's the same as the LGWebOS binding where I get the idea from. 

This only proves that there's also a hack in the LGWebOS binding and that once hacks are allowed to pass reviews, they tend to proliferate as people refer to them and take it as an example. 😎 

> For me it's valid to take this constraint and mention it e.g. in readme instead of creating a complete different concept 

I have to admit that there isn't a straight-forward way to do it in place and it is unlikely that the efforts will be spent anytime soon. So I agree with you: Let us consider it a "rare exceptional use case" that slightly misuses current concepts in order to make it work for the user and not block your tremendous work and progress on this binding.

Reply:
I did investigate a bit more and found the [oh-repeater component](https://www.openhab.org/docs/ui/components/oh-repeater.html) that allows to display dynamic tables based on item's stateOptions. I'll build a prototype to check out how this can be used as an alternative to the current concept that does not rely on dynamic dependency of channel-states being used as ui-behavioral element.